include/yaml-cpp/node/convert.h

```diff
@@ -161,83 +167,272 @@ struct convert<bool> {
+#endif
+      return *this;
+  }
+
+  push_back_iterator& operator*()     { return *this; }
+  push_back_iterator& operator++()    { return *this; }
+  push_back_iterator  operator++(int) { return *this; }
+};
+
+template <class Container> 
+push_back_iterator<Container> push_backer(Container& container)
+{
+  return push_back_iterator<Container>(container);
+}
+
+template <
+  class Key, 
+  class T, 
+  class ContainerValueType = std::pair<const Key, T>
+> class node_inserter
+{
+ private:
+  Node* m_node;
+
+ public:
+  typedef void value_type;
+  typedef void difference_type;
+  typedef void pointer;
+  typedef void reference;
+  typedef std::output_iterator_tag iterator_category;
+
+  explicit node_inserter(Node& node) : m_node(&node) {}
+
+  node_inserter& operator=(const ContainerValueType& t)
+  {
+      m_node->force_insert(t.first, t.second);
+      return *this;
+  }
+
+  node_inserter& operator*()     { return *this; }
+  node_inserter& operator++()    { return *this; }
+  node_inserter  operator++(int) { return *this; }
+};
+
+template <class T> struct convert_as_sequence
+{
+  static Node encode(const T& sequence)
+  {
+    Node node(NodeType::Sequence);
+    
+    std::copy(sequence.begin(), sequence.end(), 
+      node_push_backer<typename T::value_type>(node));
+
+    return node;
+  }
+  static bool decode(const Node& node, T& sequence)
+  {
+    if (!node.IsSequence())
+      return false;
+    sequence.clear();
+    std::copy(node.begin(), node.end(), push_backer(sequence));
+    return true;
+  }
+};
+
+template <
+  class T, 
+  class Key = typename T::key_type, 
+  class Value = typename T::mapped_type> 
+struct convert_as_map
+{
+  static Node encode(const T& map) {
     Node node(NodeType::Map);
-    for (typename std::map<K, V>::const_iterator it = rhs.begin();
-         it != rhs.end(); ++it)
-      node.force_insert(it->first, it->second);
+    std::copy(map.begin(), map.end(), node_inserter<Key, Value>(node));
     return node;
   }
 
-  static bool decode(const Node& node, std::map<K, V>& rhs) {
+  static bool decode(const Node& node, T& map)
+  {
+    if (!node.IsMap())
+      return false;
+    map.clear();
+    for (const auto& kv : node)
+    {
+#if defined(__GNUC__) && __GNUC__ < 4
+      // workaround for GCC 3:
+      map.emplace(kv.first.template as<Key>(), kv.second.template as<Value>());
+#else
+      map.emplace(kv.first.as<Key>(), kv.second.as<Value>());
+#endif
+    }
+    return true;
+  }
+};
+
+// A set is a map where each value is null.
+template <class T>
+struct convert_as_set
+{
+  static Node encode(const T& rhs) {
+    Node node(NodeType::Map);
+    for (const auto& item : rhs) node.force_insert(item, Null);
+    return node;
+  }
+
+  static bool decode(const Node& node, T& rhs) {
     if (!node.IsMap())
       return false;
 
     rhs.clear();
-    for (const_iterator it = node.begin(); it != node.end(); ++it)
+    for (const auto& kv : node)
```

Comment:
Right. Here is a larger problem:

In stlemitter.h, an `std::set` is serialized as a _yaml sequence_, not as a yaml map as I'm doing here. So what should it be? I'm now leaning more towards realizing an `std::set` as a yaml sequence.